drivers/rtc/rtc_ds1307.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2023 Arunmani Alagarsamy
+ * Author: Arunmani Alagarsamy  <arunmani27100@gmail.com>
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/util.h>
+
+#define DT_DRV_COMPAT maxim_ds1307
+
+LOG_MODULE_REGISTER(ds1307, CONFIG_RTC_LOG_LEVEL);
+
+#define DS1307_I2C_ADDR 0x68
+
+/* DS1307 registers */
+#define DS1307_REG_SECONDS 0x00
+#define DS1307_REG_MINUTES 0x01
+#define DS1307_REG_HOURS   0x02
+#define DS1307_REG_DAY     0x03
+#define DS1307_REG_DATE    0x04
+#define DS1307_REG_MONTH   0x05
+#define DS1307_REG_YEAR    0x06
+#define DS1307_REG_CTRL    0x07
+
+#define SECONDS_BITS GENMASK(6, 0)
+#define MINUTES_BITS GENMASK(7, 0)
+#define HOURS_BITS   GENMASK(5, 0)
+#define DATE_BITS    GENMASK(5, 0)
+#define MONTHS_BITS  GENMASK(4, 0)
+#define WEEKDAY_BITS GENMASK(2, 0)
+#define YEAR_BITS    GENMASK(7, 0)
+
+struct ds1307_config {
+	struct i2c_dt_spec i2c_bus;
+};
+
+struct ds1307_data {
+	struct k_spinlock lock;
+};
+
+static int ds1307_set_time(const struct device *dev, const struct rtc_time *tm)
+{
+	int err;
+	uint8_t regs[7];
+
+	struct ds1307_data *data = dev->data;
+	const struct ds1307_config *config = dev->config;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	LOG_DBG("set time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		tm->tm_year, tm->tm_mon, tm->tm_mday, tm->tm_wday, tm->tm_hour, tm->tm_min,
+		tm->tm_sec);
+
+	regs[0] = bin2bcd(tm->tm_sec) & SECONDS_BITS;
+	regs[1] = bin2bcd(tm->tm_min);
+	regs[2] = bin2bcd(tm->tm_hour);
+	regs[3] = bin2bcd(tm->tm_wday);
+	regs[4] = bin2bcd(tm->tm_mday);
+	regs[5] = bin2bcd(tm->tm_mon);
+	regs[6] = bin2bcd(tm->tm_year);
+
+	err = i2c_burst_write_dt(&config->i2c_bus, DS1307_REG_SECONDS, regs, sizeof(regs));
+
+	k_spin_unlock(&data->lock, key);
+
+	return err;
+}
+
+static int ds1307_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	int err;
+	uint8_t regs[7];
+
+	struct ds1307_data *data = dev->data;
+	const struct ds1307_config *config = dev->config;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	err = i2c_burst_read_dt(&config->i2c_bus, DS1307_REG_SECONDS, regs, sizeof(regs));
+	if (err != 0) {
+		goto unlock;
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & SECONDS_BITS);
+	timeptr->tm_min = bcd2bin(regs[1] & MINUTES_BITS);
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS); /* 24hr mode */
+	timeptr->tm_wday = bcd2bin(regs[3] & WEEKDAY_BITS);
+	timeptr->tm_mday = bcd2bin(regs[4] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[5] & MONTHS_BITS);
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
```

Comment:
You must initialize all fields of the `struct tm` according to the documentation here https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#c.rtc_time.tm_yday

So add
```
	timeptr->tm_yday = -1;
	timeptr->tm_isdst =  -1;
	timeptr->tm_nsec =  0;
```

Reply:
I'm fixing the test as well with this PR https://github.com/zephyrproject-rtos/zephyr/pull/64447, you don't have to wait for it to be merged or anything, just pointing it out :)